README
WELCOME
=======

The MicroGear project is an advance embedded application for unmanned
aircraft systems.  In includes a functioning open-source kalman filter
for estimating attitude and location based on gps and inertial
sensors.  It also includes an advance PID-based autopilot and a host
of supporting libraries and cool functionality.  It leverages "easy
XML" for config file parsing and several structures and libraries from
the FlightGear project including the "property system" and a number
advanced math routines.

MicroGear is a UAV guidance, navigation, and control device
application that is robust, highly capable, highly configurable, open,
and interfaceable.  Along with high levels of capability, MicroGear
focuses on robustness, high performance, and minimal memory and disk
footprints.


CONFIGURE/COMPILING
===================

1. Add the location of your arm cross compiler binaries to your path.
   This will probably look something like:

   export PATH=${PATH}:/home/curt/Projects/GumStix/gumstix-buildroot/build_arm_nofpu/staging_dir/bin
   export PATH=${PATH}:/usr/local/arm/3.3.2/bin

   Not that there are some "setpath-*.sh" scripts in the toplevel
   MicroGear tree.  These will need to be modified for your local
   paths, but are there as an an example and for your convenience.

2. Run "./autogen.sh"

3. Run "./configure CFLAGS="-Wall -O2" CXXFLAGS="-Wall -O2" --host=something-something"

   For the "new" Gumstix Open Embedded tools you would use
   --host=arm-angstrom-linux-gnueabi

   For the "old" gumstix/verdex tree you would add the cross compilers
   to your path and use --host=arm-linux

4. Run "make"

5. The resulting applications are called "ugear" and "decoder".  You will
   find them in src/main/ and they will need to be copied over to the
   target computer.


MAGNETOMETER HARD IRON CALIBRATION
==================================

Note that many calibration parameters are "hard" coded in the source.
You will need to rerun "make" after any changes to the code.

Hard iron calibration parameters are set in ahrs.cpp, bBx, bBy, sfx, and sfy.

Users need to fill out proper values for their unit if necessary (most
likely will be necessary for you.)  The following link explains some
of the underlying issues with magnetometer sensors as well as how to
more effectively use them.

    http://www.ssec.honeywell.com/magnetic/datasheets/amr.pdf

Calibration procedure overview:

Rotate the sensor at least one full revolution recording the sequence
of hx, hy values (see --log-file on option.)  Plot the recorded hx,hy
pairs and the data will form an ellipsoid, probably offset from the
origin.

The goal of the calibration procedure is to come up with horizontal &
vertical offset and scaling factors that transform this offset
ellipsoid back into a radius=1 circle centered on the origin.

To do this, bBx, bBy are assigned the center poing of the ellipsoid
(offset from the origin.)

sfx, sfy are vertical and horizontal scaling factors that are
multiplied against the recentered ellipsoid to stretch/squash it back
into a r=1 circle.

Any time the sensor is remounted, repositioned, or reoriented, the sensor
will need to be recalibrated.
 
The default values in the code are for a situation where no hard iron
calibration affects observed and no offset or rescaling is needed.

    #define bBy 0.0
    #define bBx 0.0
    #define sfx 1.0
    #define sfy 1.1

As an example, we mount the sensor in the passenger foot well of a VW
Jetta.  We drive two circles, plot hx, hy and observe the resulting
ellipsoid.  In this case the ellipsoid is centered at 0.25, 0.65.  The
total horizontal radius of the ellipsoid is about 0.5 and the vertical
radius is about 0.375.  The both need to be scaled to 1.0 so sfx is
set to 1/0.5 and sfy is set to 1/0.357

    #define bBy 0.65
    #define bBx 0.25
    #define sfx (1.0 / 0.500)
    #define sfy (1.0 / 0.375)



NEWS
# aura-core development summary

## Overview

Aura-core was born out of the Avior autopilot effort.  The avior is
developed by Airborne Technologies, Inc. out of Wasilla, AK.  The
original Avior flew in a small marinized (water proof) UAS and
development was funded through several NOAA.gov projects.  The Avior
development traces it's roots back to 2005-2006 and early prototypes
were flying in 2007.

The Avior system incorporated significant amount of open-source
derived code (much of it from the FlightGear open-source flight
simulator.)  Thus prior to distribution and further development, the
open-source portions of the avior were split out into their own
library and licensing was ensured to be LGPL (or more relaxed as
required) for all included modules.  Aura-core contains almost
everything needed to construct a standalone autopilot, so background
efforts have been in place through 2015 and into 2016 to fill in the
missing pieces.  As part of this, aura-core has embraced python, and
future versions will include more and more python integration, right
on board the aircraft.

## 2016 Development

* 2016-01-23: Integrate Python-based property system, paving the way
  for future python integration.

* 2016-01-18: Standardize on .cxx and .hxx for C++ source file name
  extensions

* 2016-01-15: Replace fgprops with new pyprops.  Replace C-based xml
  parser code with python xml parser.

* 2016-01-01: Work begins on replacing the original fgprops subsystem
  with a new python-based property system, exposing the entire
  property tree to native python code.

## 2015 Development

* 2015-12-14: Perform quite a bit of internal renaming for consistency
  and for the future.

* 2015-12-14: Remove the original ardupilot sensor head interface.
  This was the old/original Atmega328 based board that could drive 2
  servos accurately and 2 additional servos very coarsely.  This
  little board that did almost nothing enabled me to move beyond the
  original micronav and into a future world.

* 2015-10-06: Begin work on Goldy2 driver.  This is a Univerity of
  Minnesta UAS lab developed system that also runs on a beaglebone.
  The sensor head communicates via ethernet.

* 2015-09-16: While not flying, continuously update our ground
  altitude estimate based on current altitude.

* 2015-09-15: Remove remnants of old MNAV code/support.  Remove
  support for ancient sparkfun 6DOF imu sensor.  And a 'recalibrate'
  function fo resetting airspeed zero point.

* 2015-07-16: Add a new apm2-sensors config info message so we can log
  the apm2 serial number (for later keeping track of what flight log
  data came from which specific sensor.)

* 2015-07-15: Set system clock from gps time when gps time becomes
  valid (but only if substantially differnt from system time.)

* 2015-06-25: Make circle manager a standalone subsystem within
  aura-core.

* 2015-05-04: Log IMU temperature along with main IMU sensor data.
  EKF already estimates sensor biases (for gyros and accels).
  Developed a system to correlate EKF bias estimates with temperature
  so that each individual system can build up a temperature
  calibration fit over time as the system is flown in a variety of
  conditions.

* 2015-04-29: Add NavPy scripts to the distribution from UMN UAS lab.
  Allows rerunning all the flight data through the EKF off line and
  restimating attitude and sensor biases.  Good for offline processing
  and other types of testing and developing.

* 2015-04-22: Switch to quaternion based EKF implementation.

* 2015-02-15: Massive changes to apm2-sensor head to implement a
  'smart receiver' type system.  Mixing modes happen right on the
  aircraft (little processeor) side, not in the transmitter and not
  then replicated by the autopilot.  This allows implementing a
  universal stability/dampening system.

## 2014 (and previous)



2009-04-13: version 1.6

* Check point release prior to integrating UMN INS/GNS code and testing a 
  timer/callback architecture.


src/imugps.c
/******************************************************************************
* FILE: imugps.c
* DESCRIPTION:
*   
*   
*
* SOURCE: 
* LAST REVISED: 4/05/06 Jung Soon Jang
******************************************************************************/
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <termios.h>
#include <math.h>
#include <pthread.h>

#include "imugps.h"
#include "misc.h"
#include "serial.h"
#include "globaldefs.h"

//
// uNAV packet length definition
//
#define FULL_PACKET_LENGTH	38
#define SENSOR_PACKET_LENGTH    51   
#define GPS_PACKET_LENGTH	35
#define FULL_PACKET_SIZE        86  // scaled mode with sampling less than 100Hz
#define fullspeed		0

#define D2R			0.017453292519940
#define R2D			57.29577951308232
#define g			9.81

//temperature compensation for accel.
//temperature compensation for mag.

//
// prototype definition
//
int  checksum(byte* buffer, int packet_len);
void decode_imupacket(struct imu *data, byte* buffer);
void decode_gpspacket(struct gps *data, byte* buffer);

//
// global variables
//
extern short log_to_file;
extern char  buf_err[50];
int   	     sPort0;

struct servo servopacket;
short        autopilot_enable = FALSE;
short        control_init = FALSE;
char         *cnt_status;

// Main IMU/GPS data aquisition thread.
//
// Note this thread runs as fast as IMU/GPS data is available.  It
// blocks on the serial port read, otherwise it runs full bore,
// pulling in data as fast as it's available.
//
void *imugps_acq(void *thread_id)
{
    int		count=0,nbytes=0,headerOK=0;
    short		i=0;
    static int    GPS_INIT=FALSE,GPS_FULL=0,err_cnt=0;
    byte  	input_buffer[FULL_PACKET_SIZE]={0,};
    byte  	SCALED_MODE[11] ={0x55,0x55,0x53,0x46,0x01,0x00,0x03,0x00, 'S',0x00,0xF0};
    byte          CH_BAUD[11]     ={0x55,0x55,0x57,0x46,0x01,0x00,0x02,0x00,0x03,0x00,0xA3};
    byte		CH_SAMP[11]     ={0x55,0x55,0x53,0x46,0x01,0x00,0x01,0x00,0x02,0x00,0x9D};
    byte          CH_SERVO[7]     ={0x55,0x55,0x53,0x50,0x08,0x00,0xAB};
    byte		temp;
    FILE   	*fimu,*fgps;
  
    /*********************************************************************
     *Open Files
     *********************************************************************/
    if (log_to_file) {
        if((fimu = fopen("/mnt/cf1/imu.dat","w+b"))==NULL) {
            printf("imu.dat cannot be created in /mnt/cf1 directory...error!\n");
            _exit(-1);
        }
        if((fgps = fopen("/mnt/cf1/gps.dat","w+b"))==NULL) {
            printf("gps.dat cannot be created in /mnt/cf1 directory...error!\n");
            _exit(-1);
        }
    }

#ifndef NCURSE_DISPLAY_OPTION
    printf("[imugps_acq]::thread[%d] initiated...\n",thread_id);
#endif
  
    /*********************************************************************
     *Open and configure Serial Port2 (com2)
     *********************************************************************/
    sPort0 = open_serial(SERIAL_PORT2,BAUDRATE_38400); 
      
    while (nbytes != 11) nbytes = write(sPort0,(char*)CH_BAUD, 11);     nbytes = 0;  
    close(sPort0);
    sPort0 = open_serial(SERIAL_PORT2,BAUDRATE_57600); 
  
    
    while (nbytes != 11) nbytes = write(sPort0,(char*)CH_SAMP, 11);     nbytes = 0;  
    while (nbytes != 11) nbytes = write(sPort0,(char*)SCALED_MODE, 11); nbytes = 0;
    while (nbytes !=  7) nbytes = write(sPort0,(char*)CH_SERVO, 7);     nbytes = 0;  
  
    while (1) {
        /*********************************************************************
         *Find start of packet: the heade r (2 bytes) starts with 0x5555
         *********************************************************************/
  
        while (headerOK !=2)
            {
                while(1!=read(sPort0,input_buffer,1));
                if (input_buffer[0] == 0x55) headerOK++;
                else		 	  headerOK = 0;
            }
     	
        headerOK = 0; while(1!=read(sPort0,&input_buffer[2],1));
        nbytes = 3; 
  
        // Read packet contents
        switch (input_buffer[2]) {
        case 'S':               // IMU packet without GPS
            while ( nbytes < SENSOR_PACKET_LENGTH ) {
                nbytes += read(sPort0, input_buffer+nbytes,
                               SENSOR_PACKET_LENGTH-nbytes); 
            }

            // check checksum
            if ( checksum(input_buffer,SENSOR_PACKET_LENGTH) == TRUE ) {
                pthread_mutex_lock(&mutex_imu);
                decode_imupacket(&imupacket, input_buffer);
                pthread_cond_signal(&trigger_ahrs);
                if(log_to_file) fwrite(&imupacket, sizeof(struct imu),1,fimu);
                pthread_mutex_unlock(&mutex_imu);  
            } else {
#ifndef NCURSE_DISPLAY_OPTION 
                printf("[imu]:checksum error...!\n"); 
#endif                  	
                imupacket.err_type = checksum_err; 
            };
       		  
		  
            break;
        case 'N':               // IMU packet with GPS
            while ( nbytes < FULL_PACKET_SIZE ) {
                nbytes += read(sPort0, input_buffer+nbytes,
                               FULL_PACKET_SIZE-nbytes); 
            }

            // check checksum
            if ( checksum(input_buffer,FULL_PACKET_SIZE) == TRUE ) {
                pthread_mutex_lock(&mutex_imu);
                decode_imupacket(&imupacket, input_buffer);
                pthread_cond_signal(&trigger_ahrs);
                if(log_to_file) fwrite(&imupacket, sizeof(struct imu),1,fimu);
                pthread_mutex_unlock(&mutex_imu);  
		     
                // check GPS data packet
                if(input_buffer[33]=='G') {
                    pthread_mutex_lock(&mutex_gps);
                    decode_gpspacket(&gpspacket, input_buffer);
                    if(log_to_file) fwrite(&gpspacket, sizeof(struct gps),1,fgps);
                    pthread_mutex_unlock(&mutex_gps);
                } else {
#ifndef NCURSE_DISPLAY_OPTION		     	
                    printf("[gps]:data error...!\n");
#endif		         
                    gpspacket.err_type = got_invalid;
                } // end if(checksum(input_buffer...
            } else { 
#ifndef NCURSE_DISPLAY_OPTION		  	
                printf("[imu]:checksum error(gps)...!\n");
#endif
                gpspacket.err_type = checksum_err;
                imupacket.err_type = checksum_err; 
            }
		  
                  
            break;
        default : 
#ifdef NCURSE_DISPLAY_OPTION        
            sprintf(buf_err,"Invalid [imu] data packet (%d)",++err_cnt);
#else
            printf("[imu]:invalid data packet...!\n");
#endif		  
		  
        } // end case


        //////////////////////////////////////////////////////////////
        // NOTICE: MANUAL OVERRIDE
        //////////////////////////////////////////////////////////////

        // Manual override functionality for the uNAV is hardwired
        // into the unit.  It is hardwired to the gear channel (CH5
        // when counting from 1.)  This is not optional.  When you
        // enter manual override mode, the servo commands you send to
        // the uNAV are ignored and the unit passes through the
        // transmited values directly.
        //
        // So all the code needs to do is monitor the manual override
        // switch and not send autopilot commands in manual override
        // mode (they would be ignored anyway.)  There is no need to
        // copy the data through in software, it should all happen
        // automatically.
        //
        // Also note that I call this a "manual override" and not a
        // "fail safe".  If the uNAV fails for any reason, you will be
        // picking up toothpicks.
        //

        if ( servopacket.chn[4] <= 12000 ) {
            // if the autopilot is enabled, or signal is lost
            if ( autopilot_enable == FALSE )
                printf("[CONTROL]: switching to autopilot\n");
            autopilot_enable = TRUE;  
            count  = 15;
            cnt_status = "MNAV in AutoPilot Mode";
        } else if ( servopacket.chn[4] > 12000
                    && servopacket.chn[4] < 60000 )
        {
            // add delay on control trigger to minimize mode confusion
            // caused by the transmitter power off
            if ( count < 0 ) {
                if ( autopilot_enable == TRUE )
                    printf("[CONTROL]: switching to manual pass through\n");
                autopilot_enable = FALSE;
                control_init = FALSE;
                cnt_status = "MNAV in Manual Mode";
            } else {
                count--;
            }

        }
		
    } // end while

    //close the serial port
    close(sPort0);

    //close files
    fclose(fimu);
    fclose(fgps);

    //exit the thread
    pthread_exit(NULL);

} // end void *imugps_acq()


//
// check the checksum of the data packet
//
int checksum( byte* buffer, int packet_len )
{
    word i=0,rcvchecksum=0;
    word sum=0;

    for ( i = 2; i < packet_len - 2; i++ ) sum = sum + buffer[i];
    rcvchecksum = ((rcvchecksum = buffer[packet_len-2]) << 8) | buffer[packet_len-1];

    // if (rcvchecksum == sum%0x10000)
    if ( rcvchecksum == sum ) //&0xFFFF)
	return TRUE;
    else
 	return FALSE;
}


//
// decode the gps data packet
//
void decode_gpspacket( struct gps *data, byte* buffer )
{
    signed long tmp = 0;

    // gps velocity in m/s
    data->vn =(double)((((((tmp = (signed char)buffer[37]<<8)|buffer[36])<<8)|buffer[35])<<8)|buffer[34])*1.0e-2; tmp=0;
    data->ve =(double)((((((tmp = (signed char)buffer[41]<<8)|buffer[40])<<8)|buffer[39])<<8)|buffer[38])*1.0e-2; tmp=0;
    data->vd =(double)((((((tmp = (signed char)buffer[45]<<8)|buffer[44])<<8)|buffer[43])<<8)|buffer[42])*1.0e-2; tmp=0;

    // gps position
    data->lon=(double)((((((tmp = (signed char)buffer[49]<<8)|buffer[48])<<8)|buffer[47])<<8)|buffer[46])*1.0e-7; tmp=0;
    data->lat=(double)((((((tmp = (signed char)buffer[53]<<8)|buffer[52])<<8)|buffer[51])<<8)|buffer[50])*1.0e-7; tmp=0;
    data->alt=(double)((((((tmp = (signed char)buffer[57]<<8)|buffer[56])<<8)|buffer[55])<<8)|buffer[54])*1.0e-3; tmp=0;
   
   
    // gps time
    data->ITOW = ((data->ITOW = buffer[59]) << 8)|buffer[58];
    data->err_type = TRUE;
    data->time = get_Time();
}


//
// decode the imu data packet
//
void decode_imupacket( struct imu *data, byte* buffer )
{
    signed short tmp = 0;
    unsigned short tmpr = 0;

    /* acceleration in m/s^2 */
    data->ax = (double)(((tmp = (signed char)buffer[ 3])<<8)|buffer[ 4])*5.98755e-04; tmp=0;
    data->ay = (double)(((tmp = (signed char)buffer[ 5])<<8)|buffer[ 6])*5.98755e-04; tmp=0;
    data->az = (double)(((tmp = (signed char)buffer[ 7])<<8)|buffer[ 8])*5.98755e-04; tmp=0;
   
  
    /* angular rate in rad/s */
    data->p  = (double)(((tmp = (signed char)buffer[ 9])<<8)|buffer[10])*1.06526e-04; tmp=0;
    data->q  = (double)(((tmp = (signed char)buffer[11])<<8)|buffer[12])*1.06526e-04; tmp=0;
    data->r  = (double)(((tmp = (signed char)buffer[13])<<8)|buffer[14])*1.06526e-04; tmp=0;
   
    /* magnetic field in Gauss */
    data->hx = (double)(((tmp = (signed char)buffer[15])<<8)|buffer[16])*6.10352e-05; tmp=0;
    data->hy = (double)(((tmp = (signed char)buffer[17])<<8)|buffer[18])*6.10352e-05; tmp=0;
    data->hz = (double)(((tmp = (signed char)buffer[19])<<8)|buffer[20])*6.10352e-05; tmp=0;

    /* temperature in Celcius */
    /*
      data->Tx = (double)(((tmp = (signed char)buffer[21])<<8)|buffer[22])*6.10352e-03; tmp=0;
      data->Ty = (double)(((tmp = (signed char)buffer[23])<<8)|buffer[24])*6.10352e-03; tmp=0;
      data->Tz = (double)(((tmp = (signed char)buffer[25])<<8)|buffer[26])*6.10352e-03; tmp=0;
    */
   
    /* pressure in m and m/s */
    data->Ps = (double)(((tmp = (signed char)buffer[27])<<8)|buffer[28])*3.05176e-01; tmp=0;
    data->Pt = (double)(((tmp = (signed char)buffer[29])<<8)|buffer[30])*2.44141e-03; tmp=0;

    // servo packet
    switch (buffer[2]) {
    case 'S' :   servopacket.status = buffer[32];
        servopacket.chn[0] = ((tmpr = buffer[33]) << 8)|buffer[34]; tmpr = 0;
        servopacket.chn[1] = ((tmpr = buffer[35]) << 8)|buffer[36]; tmpr = 0;
        servopacket.chn[2] = ((tmpr = buffer[37]) << 8)|buffer[38]; tmpr = 0;
        servopacket.chn[3] = ((tmpr = buffer[39]) << 8)|buffer[40]; tmpr = 0;
        servopacket.chn[4] = ((tmpr = buffer[41]) << 8)|buffer[42]; tmpr = 0;
        servopacket.chn[5] = ((tmpr = buffer[43]) << 8)|buffer[44]; tmpr = 0;
        servopacket.chn[6] = ((tmpr = buffer[45]) << 8)|buffer[46]; tmpr = 0;
        servopacket.chn[7] = ((tmpr = buffer[47]) << 8)|buffer[48]; 
        break;
    case 'N' :   servopacket.status = buffer[67];
        servopacket.chn[0] = ((tmpr = buffer[68]) << 8)|buffer[69]; tmpr = 0;
        servopacket.chn[1] = ((tmpr = buffer[70]) << 8)|buffer[71]; tmpr = 0;
        servopacket.chn[2] = ((tmpr = buffer[72]) << 8)|buffer[73]; tmpr = 0;
        servopacket.chn[3] = ((tmpr = buffer[74]) << 8)|buffer[75]; tmpr = 0;
        servopacket.chn[4] = ((tmpr = buffer[76]) << 8)|buffer[77]; tmpr = 0;
        servopacket.chn[5] = ((tmpr = buffer[78]) << 8)|buffer[79]; tmpr = 0;
        servopacket.chn[6] = ((tmpr = buffer[80]) << 8)|buffer[81]; tmpr = 0;
        servopacket.chn[7] = ((tmpr = buffer[82]) << 8)|buffer[83]; 
        break;
    default  :
        printf("[imu]:fail to decode servo packet..!\n");
    }

    // printf("servo: %d %d %d %d %d\n", servopacket.chn[0], servopacket.chn[1], servopacket.chn[2], servopacket.chn[3], servopacket.chn[4] );
  
    data->time = get_Time();
    data->err_type = no_error;
}


void send_servo_cmd(word cnt_cmd[9])
{
    // cnt_cmd[1] = ch1:elevator
    // cnt_cmd[0] = ch0:aileron
    // cnt_cmd[2] = ch2:throttle

    byte data[24];
    short i = 0, nbytes = 0;
    word sum = 0;

    // printf("sending servo data ");
    // for ( i = 0; i < 9; ++i ) printf("%d ", cnt_cmd[i]);
    // printf("\n");

    data[0] = 0x55; 
    data[1] = 0x55;
    data[2] = 0x53;
    data[3] = 0x53;

    for ( i = 0; i < 9; ++i ) {
        data[4+2*i] = (byte)(cnt_cmd[i] >> 8); 
        data[5+2*i] = (byte)cnt_cmd[i];
    }

    // aileron
    // data[4] = (byte)(cnt_cmd[0] >> 8); 
    // data[5] = (byte)cnt_cmd[0];

    // elevator
    // data[6] = (byte)(cnt_cmd[1] >> 8);
    // data[7] = (byte)cnt_cmd[1];

    // throttle
    // data[8] = (byte)(cnt_cmd[2] >> 8);
    // data[9] = (byte)cnt_cmd[2];

    //checksum: need to be verified
    sum = 0xa6;
    for (i = 4; i < 22; i++) sum += data[i];
  
    data[22] = (byte)(sum >> 8);
    data[23] = (byte)sum;

    //sendout the command packet
    while (nbytes != 24) {
        // printf("  writing servos ...\n");
        nbytes = write(sPort0,(char*)data, 24);
    }
}
